src/wasm.h

```diff
@@ -1003,6 +1007,53 @@ class Pop : public SpecificExpression<Expression::PopId> {
   Pop(MixedArena& allocator) {}
 };
 
+class Try : public SpecificExpression<Expression::TryId> {
+public:
+  Try(MixedArena& allocator) {}
+
+  Name name;
+  Expression* body;
+  Expression* catchBody;
+
+  void finalize();
+  void finalize(Type type_);
+};
+
+class Throw : public SpecificExpression<Expression::ThrowId> {
+public:
+  Throw(MixedArena& allocator) : operands(allocator) {}
+
+  Name event;
+  ExpressionList operands;
+
+  void finalize();
+};
+
+class Rethrow : public SpecificExpression<Expression::RethrowId> {
+public:
+  Rethrow(MixedArena& allocator) {}
+
+  Expression* exnref = nullptr;
+
+  void finalize();
+};
+
+class BrOnExn : public SpecificExpression<Expression::BrOnExnId> {
+public:
+  BrOnExn() { type = unreachable; }
+  BrOnExn(MixedArena& allocator) : BrOnExn() {}
+
+  Name name;
+  Name event;
+  Expression* exnref = nullptr;
+  // This is duplicate info of param types stored in Event, but this is required
+  // when BrOnExn expression is refinalized independently without modules.
+  std::vector<Type> eventParams;
```

Comment:
`ReFinalize` needs it, and it is sometimes instantiated without a module.

Reply:
I think ReFinalize can do something like what it does for Blocks, it uses the existing type?

For instantiating it without a module, I think it can be handled like we handle Call, where the person calling `makeCall` must provide the return type?

Reply:
I think we can do that, but, unlike `block`, we are not really _refinalizing_ br_on_exn in this case, because we just don't have the info, no? And we also have to differentiate the two cases, when we have a module and we don't: so if we have a module, we gets the event type from there, and if we don't, we just leave the type as is.

The size of this array will be most likely 1 (for a single i32 pointer), so I don't think the cost of duplication will be high though. I think this will be ok, but I'm not very opinionated either. Please let me know if you think it's better to remove this.

Reply:
Talked with @kripken offline and decided to keep the copy. (Because we were gonna change the `BrOnExn`to type from event type to `exnref` type, there should be a way to know the event type even when there is no module)